File: src/main/java/xyz/phanta/cbtweaker/multiblock/MultiBlockType.java
package xyz.phanta.cbtweaker.multiblock;

import net.minecraftforge.fml.common.registry.ForgeRegistries;
import xyz.phanta.cbtweaker.CbtMod;
import xyz.phanta.cbtweaker.common.BlockMaterial;
import xyz.phanta.cbtweaker.common.CraftingBlockType;
import xyz.phanta.cbtweaker.gui.inventory.UiLayout;
import xyz.phanta.cbtweaker.recipe.RecipeLogic;
import xyz.phanta.cbtweaker.structure.StructureMatcher;

import javax.annotation.Nullable;
import java.nio.file.Path;
import java.util.Objects;

public class MultiBlockType<R, C, D> implements CraftingBlockType<R, C, D> {

    private final String id;
    private final BlockMaterial blockMaterial;
    private final Path recipeDir;
    private final RecipeLogic<R, C, D, ?, ?> recipeLogic;
    private final UiLayout uiLayout;

    // multiblock type has to be initialized at pre-init time so that the controller block can be constructed
    // however, these properties may rely on registry entries from other mods, so they must be initialized at init time
    // therefore, we use the slightly awkward late-init pattern
    @SuppressWarnings("NotNullFieldNotInitialized")
    private StructureMatcher structMatcher;
    @SuppressWarnings("NotNullFieldNotInitialized")
    private C recipeConfig;
    @SuppressWarnings("NotNullFieldNotInitialized")
    private D recipeDb;

    @Nullable
    private MultiBlockControllerBlock controllerBlock = null;

    public MultiBlockType(String id, BlockMaterial blockMaterial,
                          Path recipeDir, RecipeLogic<R, C, D, ?, ?> recipeLogic, UiLayout uiLayout) {
        this.id = id;
        this.blockMaterial = blockMaterial;
        this.recipeDir = recipeDir;
        this.recipeLogic = recipeLogic;
        this.uiLayout = uiLayout;
    }

    void init(StructureMatcher structMatcher, C recipeConfig, D recipeDb) {
        this.structMatcher = structMatcher;
        this.recipeConfig = recipeConfig;
        this.recipeDb = recipeDb;
    }

    public String getId() {
        return id;
    }

    public BlockMaterial getBlockMaterial() {
        return blockMaterial;
    }

    public Path getRecipeDirectory() {
        return recipeDir;
    }

    public StructureMatcher getStructureMatcher() {
        return structMatcher;
    }

    @Override
    public RecipeLogic<R, C, D, ?, ?> getRecipeLogic() {
        return recipeLogic;
    }

    public UiLayout getUiLayout() {
        return uiLayout;
    }

    @Override
    public C getRecipeConfig() {
        return recipeConfig;
    }

    @Override
    public D getRecipeDatabase() {
        return recipeDb;
    }

    public MultiBlockControllerBlock getControllerBlock() {
        if (controllerBlock == null) {
            controllerBlock = Objects.requireNonNull((MultiBlockControllerBlock)ForgeRegistries.BLOCKS.getValue(
                    CbtMod.INSTANCE.newResourceLocation("mb_" + id)));
        }
        return controllerBlock;
    }

    public String getTranslationKey() {
        return CbtMod.MOD_ID + ".multiblock." + id + ".name";
    }

}

File: src/main/kotlin/st/evening/mc/cbtweaker/multiblock/MultiBlockType.kt
package st.evening.mc.cbtweaker.multiblock

import net.minecraft.block.Block
import st.evening.mc.cbtweaker.CbtConsts
import st.evening.mc.cbtweaker.behaviour.MachineBehaviour
import st.evening.mc.cbtweaker.behaviour.MachineStateFactory
import st.evening.mc.cbtweaker.common.BlockConfig
import st.evening.mc.cbtweaker.common.CraftingBlockType
import st.evening.mc.cbtweaker.gui.inventory.WindowConfig
import st.evening.mc.cbtweaker.structure.StructureMatcher
import st.evening.mc.prelude.api.registration.ModRegistrar
import st.evening.mc.prelude.api.registration.block

class MultiBlockType<S>(
    reg: ModRegistrar,
    override val id: String,
    override val blockConfig: BlockConfig,
    override val behaviour: MachineBehaviour<S>,
    override val windowConfig: WindowConfig
) : CraftingBlockType<S> {
    // multiblock type has to be initialized at pre-init time so that the controller block can be constructed
    // however, these properties may rely on registry entries from other mods, so they must be initialized at init time
    lateinit var structureMatcher: StructureMatcher
        private set
    override lateinit var stateFactory: MachineStateFactory<S>
        private set

    val controllerBlock: MultiBlockControllerBlock by reg.block("mb_$id") { MultiBlockControllerBlock(this) }

    override val craftingBlock: Block
        get() = controllerBlock

    internal fun init(structureMatcher: StructureMatcher, stateFactory: MachineStateFactory<S>) {
        this.structureMatcher = structureMatcher
        this.stateFactory = stateFactory
    }

    val translationKey: String
        get() = "${CbtConsts.MOD_ID}.multiblock.$id.name"
}
